tests/conftest.py

```diff
@@ -64,34 +66,51 @@ def cognite_client_approval() -> CogniteClient:
 
     """
     with monkeypatch_cognite_client() as client:
-        state: dict[str, CogniteResourceList] = {}
-        client.iam.groups = create_mock_api(GroupsAPI, GroupList, state)
-        client.data_sets = create_mock_api(DataSetsAPI, DataSetList, state)
-        client.time_series = create_mock_api(TimeSeriesAPI, TimeSeriesList, state)
-        client.raw.databases = create_mock_api(RawDatabasesAPI, DatabaseList, state)
-        client.transformations = create_mock_api(TransformationsAPI, TransformationList, state)
+        written_resources: dict[str, Sequence[CogniteResource | dict[str, Any]]] = {}
+        deleted_resources: dict[str, list[str | int | dict[str, Any]]] = defaultdict(list)
+        client.iam.groups = create_mock_api(GroupsAPI, GroupList, written_resources, deleted_resources)
+        client.data_sets = create_mock_api(DataSetsAPI, DataSetList, written_resources, deleted_resources)
+        client.time_series = create_mock_api(TimeSeriesAPI, TimeSeriesList, written_resources, deleted_resources)
+        client.raw.databases = create_mock_api(RawDatabasesAPI, DatabaseList, written_resources, deleted_resources)
+        client.transformations = create_mock_api(
+            TransformationsAPI, TransformationList, written_resources, deleted_resources
+        )
         client.transformations.schedules = create_mock_api(
-            TransformationSchedulesAPI, TransformationScheduleList, state
+            TransformationSchedulesAPI, TransformationScheduleList, written_resources, deleted_resources
+        )
+        client.data_modeling.containers = create_mock_api(
```

Comment:
How do we treat deletion of containers and spaces now with the refactoring if there are instances? Will they fail? I believe we originally had that they failed unless --delete-data flag was present, then instances were cleaned out. If we go with my suggestion to only delete data with cdf-tk clean, it makes sense to fail deletion of spaces and containers if there are instances. Our current approach is based on default write and assume nothing exists, thus we need --drop to delete before writing. I assume we will want to change the default behaviour to upsert where there are changes (by comparing yamls). Then --drop will continue to work as today: drop the config before upserting, thus upsert becomes a write.